lib/googleauth/service_account.rb

```diff
@@ -104,7 +112,7 @@ def initialize options = {}
       # authenticate instead.
       def apply! a_hash, opts = {}
         # Use the base implementation if scopes are set
-        unless scope.nil? && target_audience.nil?
+        if !enable_self_signed_jwt? && !scope.nil? || !target_audience.nil?
```

Comment:
I don't think this is right... if `enable_self_signed_jwt` is true and `scope` is not null, `super` should execute. Shouldn't it be:

```rb
if !enable_self_signed_jwt? || !scope.nil? || !target_audience.nil?
```

Reply:
Also, I find the double negatives and order of operations confusing...

Reply:
Argh, I just spent 15 minutes working and reworking the logic out for the various cases, but I'm now convinced it's correct as-is.

`enable_self_signed_jwt` actually means "the client library has determined that any scopes provided are actually default scopes, _and_ it has also determined that the endpoint is the default endpoint."

So if `enable_self_signed_jwt` is true, and `scope` is not null (i.e. there are scopes provided) it means the client is passing down scopes, but is asserting that those are default scopes and that the endpoint is the default endpoint, so it should use self-signed JWT.

In general, the logic (assuming `target_audience` is also not set) is:
* If there are no scopes (`scope.nil?`) then we must use self-signed JWT (i.e. we do not call super) because we have no other choice. The setting of `enable_self_signed_jwt` is irrelevant. There's no way to get a token without scopes.
* If there are scopes, (`!scope.nil?`) then what we do is determined by `enable_self_signed_jwt`. Thus, `&&` is appropriate.

Reply:
Yeah you're right though, all the logic and double negatives are horribly confusing. Lemme see if I can refactor this a bit.

Reply:
@bshaffer Okay I pulled out a method and reversed the logic, and tried to document the logic. How does this look?

Reply:
MUCH better